exist-core/src/main/java/org/exist/xquery/GroupByClause.java

```diff
@@ -185,11 +189,29 @@ public Sequence postEval(final Sequence seq) throws XPathException {
     public void analyze(AnalyzeContextInfo contextInfo) throws XPathException {
         contextInfo.setParent(this);
         unordered = (contextInfo.getFlags() & UNORDERED) > 0;
+
+        final Set<QName> preGroupingTupleVariables = new HashSet<>();
+        FLWORClause prevClause = getPreviousClause();
+        while (prevClause != null) {
+            rootClause = prevClause;
+            preGroupingTupleVariables.addAll(prevClause.getTupleStreamVariables());
+            prevClause = prevClause.getPreviousClause();
+        }
+
         final LocalVariable mark = context.markLocalVariables(false);
         try {
             if (groupSpecs != null) {
                 for (final GroupSpec spec : groupSpecs) {
-                    final LocalVariable groupKeyVar = new LocalVariable(spec.getKeyVarName());
+                    final QName keyVarName = spec.getKeyVarName();
+
+                    if (spec.getGroupExpression() instanceof VariableReference groupExpressionVariableRef) {
+                        if (groupExpressionVariableRef.getName().equals(keyVarName)
```

Comment:
any reason to not combine this statements?

Reply:
Purely for the purposes of making the code easier to read